src/token-service/metadata-token-service.ts

```diff
@@ -1,78 +1,395 @@
-        if (this.token) {
-            return;
+    async dispose() {
+        this.logger.trace(MetadataTokenService.Messages.trace_dispose);
+
+        if (this.disposed) {
+            throw new Error(MetadataTokenService.Messages.err_service_is_disposed);
         }
+        this.disposed = true;
 
-        let lastError = null;
+        if (this.timer) {
+            clearTimeout(this.timer);
+            delete this.timer;
+        }
+    }
+
+    private async _getToken(): Promise<string> {
+        this.logger.trace(MetadataTokenService.Messages.trace__getToken);
+
+        try {
+            if (this.tokenLastError && Date.now() < this.tokenRefreshAt) {
+                throw this.tokenLastError;
+            }
 
-        for (let i = 0; i < 5; i++) {
+            try {
+                if (!this.token || Date.now() >= this.tokenExpiresAt) {
+                    await this.initialize();
+                } else if (Date.now() >= this.tokenRefreshAt) {
+                    try {
+                        this.token = await this.fetchToken();
+                    } catch {
+                        // keep using existing token
+                        if (Date.now() >= this.tokenStartReportTTLAt) {
+                            this.logger.warn(
+                                MetadataTokenService.Messages.warn_emaining_token_TTL,
+                                new HRInterval(this.tokenExpiresAt - Date.now()),
+                            );
+                        }
+                    }
+                }
+            } catch (error) {
+                this.tokenLastError = error as Error;
+
+                throw error;
+            }
+        } finally {
+            if (this.doUpdateTokenInBackground && !this.disposed) {
+                this.setTimer();
+            }
+        }
+
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return this.token!;
+    }
+
+    private async initialize() {
+        this.logger.trace(MetadataTokenService.Messages.trace_initialize);
+
+        delete this.token;
+
+        for (let i = 1; ; i++) {
             try {
                 // eslint-disable-next-line no-await-in-loop
                 this.token = await this.fetchToken();
+
                 break;
             } catch (error) {
-                lastError = error;
+                if (i === INITIALIZE_MAX_ATTEMPTS_OF_GET_TOKEN) {
+                    throw error;
+                }
             }
+
+            // eslint-disable-next-line no-bitwise
+            const slotsCount = 1 << Math.min(i, INITIALIZE_BACKOFF_CEILING);
+            const maxDuration = slotsCount * INITIALIZE_BACKOFF_SLOT_DURATION;
+            const duration = Math.trunc(maxDuration * (1 - Math.random() * INITIALIZE_BACKOFF_UNCERTAIN_RATIO));
+
+            // eslint-disable-next-line no-await-in-loop,@typescript-eslint/no-loop-func,no-async-promise-executor
+            await new Promise(async (resolve, reject) => {
```

Comment:
Тут прям не хватает комментария для чего это искусственное ожидание тут.

Reply:
Написал следующий комментарий
            // Repeated attempts to get token in case of an error are performed at increasing intervals and adding
            // a random component - to statistically distribute the load on the IAM.